device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/exceptions/IotHubClientException.java

```diff
@@ -0,0 +1,180 @@
+                    new IotHubClientDeviceRegistrationAuthenticationException("Failed to open the multiplexing connection", e);
+
+            // Bring the exceptions map from the cause to the root level exception, so that users don't have to use
+            // fields from inner exceptions.
+            newException.setRegistrationExceptionsMap(((MultiplexingDeviceUnauthorizedException) e).getRegistrationExceptions());
+
+            return newException;
+        }
+
+        if (e instanceof UnauthorizedException
+                || e instanceof AmqpUnauthorizedAccessException
+                || e instanceof MqttUnauthorizedException
+                || e instanceof MqttBadUsernameOrPasswordException
+                || e instanceof MqttIdentifierRejectedException)
+        {
+            return new IotHubClientUnauthorizedException("Failed to open the client due to an authentication error", e, false);
+        }
+        else if (e instanceof ServerBusyException
+                || e instanceof ThrottledException
+                || e instanceof AmqpConnectionThrottledException
+                || e instanceof AmqpConnectionForcedException
+                || e instanceof MqttServerUnavailableException)
+        {
+            return new IotHubClientThrottledException("Failed to open the client due to a throttling error", e, true);
+        }
+        else if (e instanceof AmqpInternalErrorException
+                || e instanceof InternalServerErrorException)
+        {
+            return new IotHubClientInternalServerErrorException("Failed to open the client due to the service encountering an internal service error", e, true);
+        }
+        else if (e instanceof IotHubServiceException)
+        {
+            IotHubStatusCode statusCode = ((IotHubServiceException) e).getStatusCode();
+            return new IotHubClientException("Failed to open the client. Received status code " + statusCode.name(), e, e.isRetryable());
+        }
+        else if (e instanceof ProtonIOException)
+        {
+            return new IotHubClientIOException("Failed to open the client due to network issues. See inner exception for more details", e);
+        }
+        else if (e instanceof ProtocolException)
+        {
+            return new IotHubClientIOException("Failed to open the client due to network issues. See inner exception for more details", e);
+        }
+        else
+        {
+            return new IotHubClientException("Failed to open the client. See inner exception for more details", e, e.isRetryable());
+        }
+    }
+
+    public static IotHubClientException fromStatusCode(IotHubStatusCode statusCode)
+    {
+        return fromStatusCode(statusCode, "");
+    }
+
+    public static IotHubClientException fromStatusCode(IotHubStatusCode statusCode, String errorMessage)
+    {
+        boolean isRetryable = IotHubStatusCode.isRetryable(statusCode);
+        switch (statusCode)
+        {
+            case SERVER_BUSY:
```

Comment:
is this the equivelant of a 503? Is mapping it to a throttling exception the right thing to do here? seems like it can have its own exception type.